torchvision/prototype/transforms/_transform.py

```diff
@@ -37,9 +38,30 @@ def forward(self, *inputs: Any) -> Any:
 
         params = self._get_params(flat_inputs)
 
-        flat_outputs = [
-            self._transform(inpt, params) if check_type(inpt, self._transformed_types) else inpt for inpt in flat_inputs
-        ]
+        # This is a heuristic on how to deal with simple tensor inputs:
+        # 1. If we haven't seen an image yet, we transform a simple tensor
+        # 2. If we have seen an image before, so either a simple tensor, a `datapoints.Image` or a `PIL.Image.Image`,
+        #    we return simple tensors without modification.
```

Comment:
I don't know if this is clearer but another way to explain it is:

- tensors are passed-through if there's an Image in the sample
- if there's no Image, the first tensor is treated as an Image, the rest are passed-through

Reply:
> tensors are passed-through if there's an Image in the sample

As implemented right now, this is not true. Your proposal would require us to search through the complete sample first to check if there is an explicit image, i.e. `datapoints.Image` or `PIL.Image.Image`. Currently, we just check if we found an explicit image before we hit a simple tensor. Meaning, in the following example the simple tensor would be transformed, while your proposal says that it doesn't:

```python
sample = [torch.Tensor(...), datapoints.Image(...)]
transform(sample)
```

So, we need to answer if we talked somewhat past each other in https://github.com/pytorch/vision/pull/6663#discussion_r982567437. Your proposal brings extra safety (albeit in an edge case in the first place) at the cost of performance since we now have to go through the sample twice.

Reply:
Thanks for noticing those important details @pmeier 

> Your proposal brings extra safety (albeit in an edge case in the first place)

My reasoning for the behaviour I proposed was that in cases where dicts are passed as input e.g. `{"image': Image(), "random_stuff": some_tensor}`, then we have no control over the actual order of the `flat_inputs` (or do we?), and so this can easily lead to unexpected results. Perhaps this isn't something we need to worry about too much?

> at the cost of performance since we now have to go through the sample twice.

Do we expect this to be significant?

Reply:
> Do we expect this to be significant?

I've run a small benchmark and it comes down to approximately `200ns * N` where `N` is the length of the sample.

Reply:
We decided offline that we want the behavior @NicolasHug suggested in the top comment. The extra runtime cost shouldn't be significant and thus we err on the side of caution here.